lib/AST/LookupVisibleDecls.cpp

```diff
@@ -473,19 +473,70 @@ lookupVisibleMemberDeclsImpl(Type BaseTy, VisibleDeclConsumer &Consumer,
                              GenericSignatureBuilder *GSB,
                              VisitedSet &Visited);
 
-static void lookupVisibleProtocolMemberDecls(
-    Type BaseTy, ProtocolType *PT, VisibleDeclConsumer &Consumer,
-    const DeclContext *CurrDC, LookupState LS, DeclVisibilityKind Reason,
-                                             LazyResolver *TypeResolver, GenericSignatureBuilder *GSB,
-    VisitedSet &Visited) {
+class RestateFilteringConsumer : public VisibleDeclConsumer {
+  VisibleDeclConsumer &parentConsumer;
+  const Type baseTy;
+  const DeclContext *DC;
+  LazyResolver *resolver;
+  llvm::DenseSet<DeclName> foundVars;
+  llvm::DenseMap<std::tuple<DeclName, Type>, std::set<ValueDecl *>> foundFuncs;
+
+  void validate() {
+    assert(DC && baseTy && !baseTy->hasLValueType());
+  }
+public:
+  RestateFilteringConsumer(VisibleDeclConsumer &parentConsumer,
+                           Type baseTy, const DeclContext *DC,
+                           LazyResolver *resolver)
+  : parentConsumer(parentConsumer), baseTy(baseTy), DC(DC), resolver(resolver) {
+    validate();
+  }
+
+  void foundDecl(ValueDecl *VD, DeclVisibilityKind Reason) override {
+    assert(VD);
+    if (!VD->getDeclContext()->getAsProtocolOrProtocolExtensionContext()) {
+      parentConsumer.foundDecl(VD, Reason);
+      return;
+    }
+    if (resolver)
+      resolver->resolveDeclSignature(VD);
+
+    if (!VD->hasInterfaceType()) {
+      parentConsumer.foundDecl(VD, Reason);
+      return;
+    }
+    if (!VD->getInterfaceType()->is<AnyFunctionType>()) {
+      if (foundVars.insert(VD->getFullName()).second) {
+        parentConsumer.foundDecl(VD, Reason);
+      }
+      return;
+    }
+    auto type =
+      VD->getOverloadSignatureType()->getAs<AnyFunctionType>()->getResult();
+
+    if (foundFuncs[{VD->getFullName(), type}].empty()) {
+      parentConsumer.foundDecl(VD, Reason);
+    }
+    foundFuncs[{VD->getFullName(), type}].insert(VD);
```

Comment:
The return value of insert() tells you if the entry was inserted or not so you can use that instead.

Why is the value of this map a set of types here, do you actually do anything with them?

Reply:
> The return value of insert() tells you if the entry was inserted or not so you can use that instead.

It's safer to check for emptyness, because I have a set a pointers. Actually, I don't want more that one element that is equal by value, so I better change the container to `TinyVector`.